Implement heartbeats for Ruby

unflxw opened the pull request:
Use the Transmitter to send heartbeats to AppSignal. Based on @matsimitsu's implementation (see #1054)

Part of #1054. See also test setup: https://github.com/appsignal/test-setups/pull/191

### FAQ (Figuratively Asked Questions)

#### Why implement it using the transmitter?

It gives us HTTP(S) proxy support for free.

#### The implementation in #1054 uses `send_error` to send an error to AppSignal when the heartbeat is not sent. Why doesn't this implementation do so?

I believe we should use AppSignal to notify the customer of issues with their code, not of issues with AppSignal itself.

#### The implementation in #1054 stores the "current heartbeat" in `Thread.current`. Why doesn't this implementation do so?

There can be more than one heartbeat active at the same time in the same thread, which could lead to `heartbeat_finish` finishing the wrong heartbeat, and other heartbeats not being finished. (This would happen, for example, if automatically instrumenting nested Rake jobs)

Ultimately, storing it in `Thread.current` does not have any significant benefit over the following:

```ruby
heartbeat = Appsignal.Heartbeat.new()
# The customer can throw this at `Thread.current`,
# if that is useful to them!

def somewhere()
  heartbeat.start()
end

def somewhere_else()
  heartbeat.finish()
end
```

#### You mention automatically instrumenting Rack jobs. Why doesn't this implementation do so?

Primarily, because the UI for the heartbeats functionality is not (yet) designed for arbitrary heartbeats, which have not been previously declared in the UI, to be sent.

If this was to be done, the implementation on Rack's end would be simple, merely wrapping [this `super`](https://github.com/appsignal/appsignal-ruby/blob/f72df408417b304e6c0a9cbf413d58fd2997f31d/lib/appsignal/integrations/rake.rb#L7) in an `Appsignal.heartbeat(name)` block.

But the heartbeat implementation itself would need to be changed, in order to send heartbeats in "batches" every few seconds, rather than triggering an immediate HTTP request for each, as that would incur a significant overhead for quick, deeply nested jobs, an overhead that the customer would not have an easy way to opt out of.

It might be best to open the above can of worms only once, in the agent, rather than within each integration.

luismiramirez commented on lib/appsignal/heartbeat.rb:
```diff
@@ -0,0 +1,59 @@
+    @transmitter = nil
+
+    def self.transmitter
+      @transmitter ||= Appsignal::Transmitter.new(
```
As rubocop doesn't like this, you can also put this in the eigenclass and use an accessor.

tombruijn commented on lib/appsignal/heartbeat.rb:
```diff
@@ -0,0 +1,71 @@
+
+    def transmit_event(kind)
+      unless Appsignal.active?
+        Appsignal.internal_logger.debug("AppSignal not active, not transmitting heartbeat event")
```
This needs and early return if you do not want it to send the event. Can you add a test for this scenario too?

```suggestion
        Appsignal.internal_logger.debug("AppSignal not active, not transmitting heartbeat event")
        return
```

unflxw replied:
Big whoops here, thanks for noticing it. Will do.

tombruijn commented on lib/appsignal/heartbeat.rb:
```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require "net/http"
+require "json"
+
```
These two requires aren't used here and are already included in the transmitter.rb file this class relies on to send requests.

```suggestion
```

unflxw replied:
Thank you for noticing that! (I started by modifying @matsimitsu's implementation to use `Net::HTTP` instead of RestClient, and then I switched to using the processor, but I forgot to remove these imports)

tombruijn commented on lib/appsignal/heartbeat.rb:
```diff
@@ -0,0 +1,71 @@
+      response = self.class.transmitter.transmit(event(kind))
+
+      unless response.code.to_i >= 200 && response.code.to_i < 300
+        Appsignal.internal_logger.warn(
+          "Failed to transmit heartbeat event: status code was #{response.code}"
+        )
+      end
+    rescue => e
+      Appsignal.internal_logger.warn("Failed to transmit heartbeat event: #{e}")
```
As discussed in the Elixir PR, make these error level logs, because something failed to send.

```suggestion
        Appsignal.internal_logger.error(
          "Failed to transmit heartbeat event: status code was #{response.code}"
        )
      end
    rescue => e
      Appsignal.internal_logger.error("Failed to transmit heartbeat event: #{e}")
```

tombruijn commented on spec/lib/appsignal/heartbeat_spec.rb:
```diff
@@ -0,0 +1,62 @@
+
+  describe ".heartbeat" do
+    describe "when a block is given" do
+      it "should send a heartbeat start and finish and return the block output" do
```
Would be good to add a test case for when an error occurs in the block, that it doesn't send the finish event.

unflxw replied:
Ooh, yes, absolutely. Thank you for spotting that! 💯

unflxw commented:
@tombruijn Thank you so much for taking a proper look at this PR and giving me thoughtful feedback! I have now addressed the points you raised. Please re-review.

unflxw commented:
Moved this PR to Blocked. It is ready to merge any time, ideally shortly before the user-facing feature in the product is available.
